src/class-controller.php

```diff
@@ -0,0 +1,215 @@
+		$this->init();
+	}
+
+	/**
+	 * Add hooks to kick off migrator and tie in useful query vars.
+	 */
+	protected function init() {
+		// Post types and taxonomies must be registered by this point to be used
+		// for determining migratability.
+		add_action( 'setup_theme', [ $this, 'run' ], 100 );
+
+		// Add query vars for post types to our WP.
+		add_action( 'registered_post_type', [ $this, 'add_post_type_query_vars' ], 10, 2 );
+
+		// @todo Add query vars for taxonomies.
+	}
+
+	/**
+	 * Run the migrator.
+	 */
+	public function run() {
+		// Set up our extended WP instance to establish query vars early.
+		$this->set_up_wp();
+
+		// Determine migratability of the current request.
+		$this->set_migratability();
+
+		// Maybe migrate.
+		$this->maybe_migrate();
+	}
+
+	/**
+	 * Sets up WP_Theme_Migrator_WP instance and parses the request.
+	 */
+	protected function set_up_wp() {
+		$this->wp->main();
+	}
+
+	/**
+	 * Set migratability of the current request.
+	 */
+	protected function set_migratability() {
+		/**
+		 * Skip migration checks while doing cron.
+		 *
+		 * @param bool              $skip_during_cron Whether to skip.
+		 * @param WP_Theme_Migrator $this This WP_Theme_Migrator instance.
+		 */
+		$skip_during_cron = apply_filters( 'wp_theme_migrator_skip_during_cron', true, $this );
+
+		if ( $skip_during_cron && defined( 'DOING_CRON' ) && DOING_CRON ) {
+			return;
+		}
+
+		$this->should_migrate = $this->check_migratability();
+	}
+
+	/**
+	 * Migrate, if we should.
+	 */
+	protected function maybe_migrate() {
+		if ( ! $this->should_migrate ) {
+			return;
+		}
+
+		$this->switch_theme();
+
+		// Only migrate once in a request.
+		$this->stop_migrator();
+	}
+
+	/**
+	 * Switch to the new theme.
+	 */
+	protected function switch_theme() {
+		// @todo Add child theme support.
+		// @todo Add sidebar and menu support.
+
+		// Filtering the theme-related options early ensures the STYLESHEETPATH
+		// and TEMPLATEPATH constants are set.
+		add_filter( 'option_stylesheet', [ $this, 'get_theme_stylesheet' ] );
+		add_filter( 'option_template', [ $this, 'get_theme_template' ] );
+		add_filter( 'option_current_theme', [ $this, 'get_theme_name' ] );
+	}
+
+	/**
+	 * Get the theme stylesheet.
+	 *
+	 * @return string Theme slug.
+	 */
+	public function get_theme_stylesheet(): string {
+		return $this->context->get_theme_slug();
+	}
+
+	/**
+	 * Get the theme template.
+	 *
+	 * @return string Theme slug.
+	 */
+	public function get_theme_template(): string {
+		return $this->context->get_theme_slug();
+	}
+
+	/**
+	 * Get the name of the theme.
+	 *
+	 * @return string Theme name.
+	 */
+	public function get_theme_name(): string {
+		return $this->context->get_theme_name();
+	}
+
+	/**
+	 * Check if the current request is migratable.
+	 *
+	 * @return bool Whether the request is migratable.
+	 */
+	protected function check_migratability(): bool {
+		if ( empty( $this->context->get_callbacks() ) ) {
```

Comment:
If `get_callbacks()` reliably returns an array, this block can be removed, because the `foreach` will not execute on an empty array, and then it'll `return false` at the bottom of the function.